### PR TITLE
Allow binding lookup without relation ids

### PR DESCRIPTION
**Pull request by dshcherb**
* Networking information associated with a relation endpoint can be
  useful before the relations themselves exist;
* It is not possible to retrieve networking info for dead relations so
  for those cases networking info is looked up as if a relation ID was
  not specified.

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -60,9 +60,13 @@ def get_relation(self, relation_name, relation_id=None):
...
+    def get_binding(self, binding_key):
+        """Get a network space binding.
+
+        binding_key -- a name of a relation or a Relation object. Providing a Relation object will return
+                       a Binding object specific to a relation.
+        """
+        return self._bindings.get(binding_key)
 
 
 class ModelCache:
```
Comment (jameinel) [inline ops/model.py]: should we make ModelCache private (_ModelCache), I don't think it is something we actively want charmers to interact with, right?

File: ops/model.py [hunk 2]
```diff
@@ -279,7 +290,12 @@ def __init__(self, name, relation_id, backend):
     @property
     def network(self):
         if self._network is None:
-            self._network = Network(self._backend.network_get(self.name, self._relation_id))
+            try:
+                self._network = Network(self._backend.network_get(self.name, self._relation_id))
+            except RelationNotFoundError:
+                # If a relation is dead, we can still get network info associated with an endpoint itself
+                # not associated with a particular relation id.
+                self._network = Network(self._backend.network_get(self.name))
```
Comment (jameinel) [inline ops/model.py]: We might want to actually do:

if self._relation_id is not None:
  self._network = ...

Because otherwise it calls with the same args twice.

File: test/test_model.py [hunk 3]
```diff
@@ -657,6 +657,13 @@ def test_storage(self):
                 self.model.storages.request('data', count_v)
 
     def test_relation_endpoint_bindings(self):
```
Comment (jameinel) [inline test/test_model.py]: This test feels like it has grown organically into something that is now a bit hard to actually understand. I'll try to grab the whole source of it, but I wonder if we can find a structure that would be a bit simpler.

File: ops/model.py (hunk 2 above)
Comment (dshcherb) [reply ops/model.py]: https://github.com/canonical/operator/pull/157/commits/a9e15b0059086c7c3fe20af4af18e281cb70ad30

File: ops/model.py [hunk 4]
```diff
@@ -60,9 +60,13 @@ def get_relation(self, relation_name, relation_id=None):
         """
         return self.relations._get_unique(relation_name, relation_id)
 
-    def get_binding(self, relation):
-        """Get a network space binding for a Relation object."""
-        return self._bindings.get(relation)
+    def get_binding(self, binding_key):
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply ops/model.py]: `binding_key` allows us to overload it a little bit more although I don't think we will.

`relation_or_name` seems a bit long and it doesn't specify that it's actually `relation_or_relation_name`.

File: ops/model.py (hunk 1 above)
Comment (dshcherb) [reply ops/model.py]: I think I could have a separate PR for this one.

File: test/test_model.py (hunk 3 above)
Comment (jameinel) [reply test/test_model.py]: Looking at it inline makes me feel like it is a strong case for having a TestModelBindings that has its own setUp and helper methods rather than putting all of that context into local methods with closures.

File: test/test_model.py [hunk 5]
```diff
@@ -714,49 +721,70 @@ def test_relation_endpoint_bindings(self):
...
+            self.assertEqual(binding.network.interfaces[2].address, ipaddress.ip_address('192.0.3.3'))
+            self.assertEqual(binding.network.interfaces[2].subnet, ipaddress.ip_network('192.0.3.3/32'))
+            self.assertEqual(binding.network.interfaces[3].name, 'tun')
+            self.assertEqual(binding.network.interfaces[3].address, ipaddress.ip_address('2001:db8::3'))
+            self.assertEqual(binding.network.interfaces[3].subnet, ipaddress.ip_network('2001:db8::3/128'))
+            self.assertEqual(binding.network.interfaces[4].name, 'tun')
+            self.assertEqual(binding.network.interfaces[4].address, ipaddress.ip_address('fe80::1:1'))
+            self.assertEqual(binding.network.interfaces[4].subnet, ipaddress.ip_network('fe80::/64'))
+
+        single_binding_test_cases = [(
```
Comment (jameinel) [inline test/test_model.py]: This in particular is a lot of abstraction that is quite hard to follow, for which we only have 2 cases that are actually evaluated.

File: ops/model.py [hunk 6]
```diff
@@ -60,9 +60,13 @@ def get_relation(self, relation_name, relation_id=None):
...
+        """Get a network space binding.
+
+        binding_key -- a name of a relation or a Relation object. Providing a Relation object will return
```
Comment (jameinel) [inline ops/model.py]: I've usually used the Python rST docstring format:
https://www.python.org/dev/peps/pep-0287/

I have also seen things like "Google style doctrings":
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

There doesn't seem to be One True Way:
https://stackoverflow.com/questions/3898572/what-is-the-standard-python-docstring-format

But I think we should look to standardize as a project, so that someone coming to the project doesn't have to reinterpret each command. It also will let us autogenerate some pretty html docs if we have a standard format.

Comment (chipaca) [reply ops/model.py]: agreed we need to standardise. I don't think it should block this PR, as the whole codebase is a bit of a mismatch --- we should wrap up our conversation around this and land a single PR that Does The Thing.

File: ops/model.py (hunk 2 above)
Comment (jameinel) [reply ops/model.py]: This seems a bit verbose vs:
```
            try:
                self._network = Network(self._backend.network_get(self.name, self._relation_id))
            except RelationNotFoundError:
                if self._relation_id is not None:
                   # If a relation is dead, we can still get network info associated with an endpoint itself
                   self._network = Network(self._backend.network_get(self.name))
```

Comment (chipaca) [reply ops/model.py]: that's a different behaviour though, what's there now is more like
```python
            try:
                self._network = Network(self._backend.network_get(self.name, self._relation_id))
            except RelationNotFoundError:
                if self._relation_id is None:
                    raise
                # If a relation is dead, we can still get network info associated with an endpoint itself
                self._network = Network(self._backend.network_get(self.name))
```

**Comment by jameinel** [comment]
right, I did intend the extra 'raise', I missed it in my suggestion

File: ops/model.py (hunk 6 above)
Comment (dshcherb) [reply ops/model.py]: There was a discussion a while ago about that (I was suggesting this https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists to be used) but we never went along with that.

I think it might be a good time to revisit it for the whole code base but let's do the switch in a different PR when we decide on a particular approach.

File: test/test_model.py (hunk 5 above)
Comment (dshcherb) [reply test/test_model.py]: https://github.com/canonical/operator/pull/157/commits/d806429c7daeddd6d0d3c45f8646698fc015c760

File: test/test_model.py (hunk 3 above)
Comment (dshcherb) [reply test/test_model.py]: https://github.com/canonical/operator/pull/157/commits/d806429c7daeddd6d0d3c45f8646698fc015c760

File: ops/model.py (hunk 2 above)
Comment (dshcherb) [reply ops/model.py]: https://github.com/canonical/operator/pull/157/commits/d806429c7daeddd6d0d3c45f8646698fc015c760

File: ops/model.py (hunk 1 above)
Comment (dshcherb) [reply ops/model.py]: See https://github.com/canonical/operator/pull/170

File: ops/model.py (hunk 6 above)
Comment (jameinel) [reply ops/model.py]: That would also be my preferred syntax (just from the syntaxes that I've used in the past).
I definitely agree that it is irrelevant for *this* patch, but I did want to remember for us to have the discussion.

File: ops/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dshcherb) [reply ops/model.py]: I'm going to leave it as-is for now unless it's a blocker for the PR.
https://github.com/canonical/operator/pull/157/commits/d806429c7daeddd6d0d3c45f8646698fc015c760

+ @chipaca if you have an opinion.

Happy to change it if it

[the remark this answers is not in the document]
Comment (chipaca) [reply ops/model.py]: I think `relation_or_name` and `binding_key` transmit the same amount of information (not a lot) :slightly_smiling_face:. Both need help from the docstring. Given this, I'd go with the shorter one.

File: ops/model.py [hunk 7]
```diff
@@ -60,9 +60,13 @@ def get_relation(self, relation_name, relation_id=None):
...
+                       a Binding object specific to a relation.
+        """
```
Comment (chipaca) [inline ops/model.py]: I've been trying to rephrase this to be clearer three or four times now, and failing.

The docstring as it stands just adds to the confusion, and unless you know what a "Binding object specific to a relation" is, and what its ... contrapositive? is, then it's not at all clear what you get if you pass in a string.

I get that "bindings" are the juju names for these, and "extra bindings" being bindings that aren't binding anything with anything are part of the problem we're fixing longer-term in juju itself, but this conflation here doesn't seem to help.

Where is this used, such that we want a single method? Wouldn't it make more sense to have a `get_binding` and a `get_binding_by_name`? (or maybe even a `get_extra_binding` and we treat "extra binding" as a magic term that juju gives us that has little connection with the other "binding" thing juju gives us?)

Comment (niemeyer) [reply ops/model.py]: This is the case we discussed previously: we're trying to cut down on the required API for it while we don't have a crystal clear understanding of what the new version of that feature will look like, so we don't have to be obsoleting multiple entry points if possible.

Here is some suggested phrasing:
> """Get a network space binding.
> 
> binding_key -- The relation name or instance to obtain bindings for.
>
> If binding_key is a relation name the method returns the default binding for that relation. If a relation instance is provided the method first looks up a more specific binding for that specific relation ID, and if none is found falls back to the default binding for the relation name.
> """

File: ops/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (niemeyer) [reply ops/model.py]: Prefer _binding_key_ as well, unless you want to name it as _relation_instance_or_relation_name_ 

:sweat_smile:

File: ops/model.py [hunk 8]
```diff
@@ -258,12 +262,19 @@ def __init__(self, backend):
...
-        binding = self._data.get(relation)
+    def get(self, binding_key):
+        if isinstance(binding_key, Relation):
+            binding_name = binding_key.name
+            relation_id = binding_key.id
+        elif isinstance(binding_key, str):
+            binding_name = binding_key
+            relation_id = None
+        else:
+            raise ModelError('a binding key must be a str or Relation, not {}'.format(type(binding_key).__name__))
```
Comment (niemeyer) [inline ops/model.py]: Suggested message tuning:

_'binding key must be str or relation instance, not {}'_

File: ops/model.py [hunk 9]
```diff
@@ -279,7 +290,13 @@ def __init__(self, name, relation_id, backend):
     @property
     def network(self):
         if self._network is None:
-            self._network = Network(self._backend.network_get(self.name, self._relation_id))
+            try:
+                self._network = Network(self._backend.network_get(self.name, self._relation_id))
```
Comment (niemeyer) [inline ops/model.py]: Given how much we already know of the upcoming feature, we should discuss whether to go forward with Network, or rename this to something related to "spaces". Should probably catch up with the juju team on this to avoid terminology discrepancies.

File: ops/model.py (hunk 7 above)
Comment (dshcherb) [reply ops/model.py]: > returns the default binding for that relation.

`default binding for a relation` is a new term. Are we OK with introducing it here?

There is an existing term `a default binding for an application` which means that any relation endpoint will be bound to a specified space by default unless a different one is specified.

https://github.com/juju/juju/blob/27d29321e2ab68d60310d136de8daa24d1266c8a/cmd/juju/application/bind.go#L83
https://github.com/juju/juju/blob/27d29321e2ab68d60310d136de8daa24d1266c8a/core/network/space.go#L16-L20 (also named as the "alpha" space in the newer code base)

For example:

metadata:
```
provides:
    db:
        interface: db
    logging:
        interface: logging
    monitoring:
        interface: monitoring
```
bundle.yaml, bindings section of an app:
```
...
applications:
  mycharm:
    # ...
    bindings:
      "": my-default-space # this is a syntax for specifying a default binding
      db: red-space
      # both "monitoring" and "logging" endpoints will be bound to "my-default-space"
```

`get_binding('db').network.<something>` will return network info not tied to a particular relation ID but from the "red-space" (not a space provided in a default space binding for this application).

I would like to avoid overloading the term "default binding" for an app vs a relation because it is already confusing enough.

@chipaca I originally had two methods in one of the previous PRs but we decided to go for a single one for the reasons @niemeyer is outlining.

Comment (niemeyer) [reply ops/model.py]: Okay, here is some food for thought:

1. What will the method return when you use a relation name?
2. What will the method return when you use a relation instance and there was a binding to that particular relation instance?
3. What will the method return when you use a relation instance and there is a binding to the relation name but not the relation instance?

What's the proposal for naming things in these scenarios?

Comment (dshcherb) [reply ops/model.py]: Based on the definition of an application endpoint I would have "app endpoint bindings" and "relation-specific app endpoint bindings":

https://jaas.ai/docs/concepts-and-terms#heading--endpoint
> An endpoint (or application endpoint) is used to connect to another application’s endpoint in order to form a relation.

My logic is as follows:

* we can only bind app endpoints;
* a relation is formed by _connecting two endpoints_ of different apps and gets an id;
* The lookup of network info is affected by 3 factors: the **current app**'s endpoint bindings, endpoint name, relation id. This means that using a term "relation binding" is not correct because we do not bind relations themselves.

So:

1. app endpoint binding;
2. relation-specific app endpoint binding;
3. app endpoint binding.

Comment (niemeyer) [reply ops/model.py]: @dshcherb We seem to be in different wavelengths.  I don't think anyone would understand if we said "we return the app endpoint binding" in any of those cases.

File: ops/model.py (hunk 9 above)
Comment (dshcherb) [reply ops/model.py]: @jameinel @mitechie ^

File: ops/model.py (hunk 8 above)
Comment (dshcherb) [reply ops/model.py]: https://github.com/canonical/operator/pull/157/commits/cea632893deb9afe259a2dab23d6f1fd12076e0e

File: ops/model.py (hunk 7 above)
Comment (dshcherb) [reply ops/model.py]: https://github.com/canonical/operator/pull/157/commits/cea632893deb9afe259a2dab23d6f1fd12076e0e

File: ops/model.py (hunk 9 above)
Comment (chipaca) [reply ops/model.py]: This isn't a blocker for this PR though (as `Network` is already exposed)

File: test/test_model.py [hunk 10]
```diff
@@ -899,6 +893,49 @@ def test_relation_endpoint_bindings(self):
...
+        for name in (object, 0):
+            with self.assertRaises(ops.model.ModelError):
+                self.model.get_binding(name)
+
+    def test_dead_relations(self):
+        fake_script(self, 'network-get',
+                    '''[ "$1" = db0 -a "$2" = --format=json ]'''
+                    ''' && echo '{}' || '''.format(self.network_get_out) +
+                    '''{ echo ERROR invalid value "$2" for option -r: '''
+                    '''relation not found >&2 ; exit 2; }'''),
```
Comment (chipaca) [inline test/test_model.py]: how about
```suggestion
        fake_script(
            self,
            'network-get',
            '''
                if [ "$1" = db0 ] && [ "$2" = --format=json ]; then
                    echo '{}'
                else
                    echo ERROR invalid value "$2" for option -r: relation not found >&2
                    exit 2
                fi
            '''.format(self.network_get_out))
```
?
